### ISSUE TITLE
Suggestion: Raise the minimum honor requirement for authoring katas

### ISSUE THREAD

**Issue by docgunthrop** [Codewars/codewars.com#1245]
There seems to be a rising trend in low-quality katas being published. *Low-quality* meaning katas that are:
- duplicates
- poorly tested
- so simple that it's essentially a slight variation of another 8 kyu kata (which can be grouped with duplicates)

The vast majority of these are authored by relatively new members on CW (those with around 300 honor points or less).

Just today, one author with 104 honor points published seven katas in quick succession. All of them were promptly retired within 12 hours of publishing.

To continue with what was mentioned in #1118 , I suggest raising the minimum honor requirement for publishing katas in an effort to mitigate the addition of low-quality katas.

**Comment by Voileexperiments** [comment]
For the record, that incident yesterday was bad because:

- It has too much introduction, and at the same time is badly structured and has underspecified requirements, so lots of things were guesswork
- tests were lacking
- most importantly, those katas were written in a coding bootcamp oriented style, and the author doesn't realize that CW is, for the most part, a competitive programming site. (Some authors are nice enough to make the katas informative though, which is a plus.)

This particular case really has nothing to so with amount of user honor. For the record, [even 1kyu user with 12k honor can make tons of such katas).

**Comment by DonaldKellett** [comment]
@docgunthrop has a good point - if the minimum honor requirement were raised for authoring Kata then newcomers will have a better chance to study and observe the Description and Test Cases of approved Kata with a high satisfaction rating before taking a stab at authoring their own.  This would reduce the amount of disappointment for both the new Kata author and moderators reviewing the Kata when it is published.

As for the point @Voileexperiments highlighted regarding a `12k, 1 kyu` user doing the same thing as the typical newcomer, yes, such situations do occur but they are among the minority - such users will really have to not care about Codewars at all despite having used it for so long to do such a thing that clearly degrades the Codewars experience.

My suggestion on top of @docgunthrop 's (regarding raising the honor bar) is to introduce a mandatory article on what constitutes a good Kata, followed by a mandatory quiz which asks a few common sense questions such as "Choose which Kata is of the highest quality of the four: (a) Kata with misleading description, only one fixed assertion and no random assertions; (b) Kata with well-structured Description clearly separating Story from Task, has full code coverage by including ten fixed assertions and five edge assertions and contains ten random tests; (c) ... ; (d) ... ".  The new Codewarrior should be able to answer all questions correctly on the mandatory quiz before he/she is allowed to author his/her first Kata.  What do you think @jhoffner @kazk  ?

**Comment by docgunthrop** [comment]
@Voileexperiments On the contrary, there is a notable degree of correlation between a user's level of honor and the quality of katas they publish. It makes sense that a newer user, having less experience and being less "acclimated" to the CW site, would be more prone to publishing duplicates (having completed fewer katas), not implementing a proper or adequate test suite, and not following best practices.

The 1kyu user you linked is more an exception rather than the norm. I could've easily thrown in some more recent examples of poorly made katas to support my point, but refrained from doing so for the sake of courtesy.

**Comment by kazk** [comment]
@DonaldKellett
I think making users read documentations like [Kata Best Practices](https://github.com/Codewars/codewars.com/wiki/Kata-Best-Practices) and [Create Your First Kata](https://github.com/Codewars/codewars.com/wiki/Tutorial%3A-Create-Your-First-Kata) is a good idea.
I'm not sure about mandatory quizzes. It might be helpful for some users, but I don't think it'll be effective enough to justify the cost. Some reasons are:

- Knowing what's good doesn't mean s/he can author one.
- Difficult to come up with good questions to ask. It's easy to reduce the number of choices by looking at negative words like "misleading" in "Kata with misleading description..."
- Multiple choice quizzes are easy to bypass by trial and error. It can get complicated quickly if we try to prevent that.

**Comment by dinglemouse2250** [comment]
@kazk 
> I think making users read documentations like Kata Best Practices and Create Your First Kata is a good idea.

Except, I don't think you can really "make" anybody read anything. 

@DonaldKellett 

> mandatory quiz

I don't really see how there is any way to verify compliance...

~~

How about compromise and just call it a "friendly reminder" at publication time. 

Like this:

```
You are one click away from publishing this Kata...

But just before you do:
* are you aware of the Kata Best Practices wiki?
* does your Kata have example tests
* does your Kata Solution tests include have random tests
* is the Task clearly identified in the Kata Description?
* have you checked to confirm this is not a duplicate of an existing Kata?
* yada yada whatever..

If you answer YES to all of the above then proceed!
```

**Comment by kazk** [comment]
> Except, I don't think you can really "make" anybody read anything.

Yeah, actually I was thinking of something similar to your reminder. No more "this is my first..." or "didn't know...".

But those users who doesn't care won't read those reminders either.
This repository's issue template contains

> Otherwise delete all of this text and file your issue. Make sure to include steps to reproduce the issue, including which pages/kata you found the issue on, and your Codewars username if its related to your account.

and many people creates issues without removing those lines and without any useful information.


If I had unlimited resources and time, I'd write KataLint that prevents low-quality kata from publishing.

**Comment by Blind4Basics** [comment]
Hi,

Where are we on this topic? Because I just faced [that... "thing"...](https://www.codewars.com/kata/find-the-number-5). 8 kyu author, 23 honor points, inscription on October 2017. 

Something isn't working properly, so?

cheers,

B4B

**Comment by Voileexperiments** [comment]
@Blind4Basics #1120

**Comment by anter69** [comment]
I suggest to put some basic tips in the kata editor's "description" box by default:

![image](https://user-images.githubusercontent.com/12795864/37157227-8a8a8462-22e8-11e8-8d81-e0cebd47952f.png)

(taken from https://github.com/Codewars/codewars.com/wiki/Kata-Best-Practices -- it's a good base, but should be reduced even more for this purpuse)


And if that's done, go a step further, and put a skeleton test suite/example test/function in the relevant boxes, based on best practices:

![image](https://user-images.githubusercontent.com/12795864/37157375-feb914b6-22e8-11e8-9f43-e6b9a1c4a3fd.png)


Implementing these would not require much effort, and it would be available for every kata author without any further hassle (= you don't even have to click anywhere). This would hopefully prevent more people from publishing the millionth fizzbuzz kata, forgetting example tests or random tests, etc.

I believe the power users could quickly sketch these for all languages, making it even less painful to implement.

**Comment by Voileexperiments** [comment]
The problem is, there are lots of pitfalls (and anti-cheats) and we need to demonstrate them as well:

- Reference solution for random tests should go to test fixture, not preloaded, so your template needs a skeleton reference solution as well
- Python: everything in the test fixture should be put in a closure function because otherwise they're all in global scope
- JS: distinction between `Test.assertEquals`, `Test.assertSimilar` and `Test.assertDeepEquals`
- Input modification: prevent user solution mutating the input from poisoning everything else during random tests
- RNG seeding: It's mandatory in languages like C, and in lots of languages (including Python/Ruby) they're trivial to be re-seeded by user code to pin down the random tests. It's just that nobody has abused this (yet)
- Source code reading should be done before user code is even loaded/run, so user can't overwrite them before it's read by the tests. This is especially tricky in languages like C#/Java, which the anti-cheat is difficult to see (e.g in Java you need to put `@Before` attribute at the method reading user code file)
- Also, 10 random tests are definite not enough

At which point there are too many subtle details the text fixture is not big enough for all of them. And frankly, they're confusing to kata makers too.

---

Also, you should add a whole lot from https://github.com/Codewars/codewars.com/wiki/Tutorial%3A-Create-Your-First-Kata: it contains most of the *actual* stuff that results in hot beta garbage.

**Comment by anter69** [comment]
The issues you mention are much more delicate compared to the issues we see on a daily basis ("no random tests", "missing example tests", "user's result and expected are swapped", etc). So while I agree that those issues should be considered and  recorded for "serious" kata authors, putting some **basic guidelines** under the nose of **every kata author** could considerably improve the general quality of most new katas (which are usually in the 6-8 kyu range). E.g. for a usual "fizzbuzz" level kata, I wouldn't worry about RNG seeding or the global scope -- rather that it has only 5 fixed tests in total, and that it's the 20th iteration of a basic idea...

**Comment by dinglemouse2250** [comment]
Something is better than nothing. 

IMO a simple change made now is much better than debating a perfect fix which may never see light of day. Just implement my suggestion (of Dec-8 last year) and revisit this issue after a few months to see if it had desired affect.

**Comment by Voileexperiments** [comment]
Note that the honor requirement is now 300 due to #1342.

**Comment by anter69** [comment]
How about the idea of providing a sample test-suite/function/description for each language?
